src/slider/slider.tsx

```diff
@@ -0,0 +1,285 @@
+    [max, min, value, values]
+  );
+  const validStep = useMemo(() => (step < 0 ? 0 : step), [step]);
+  const isRange = useMemo(() => isArray(defaultValue ?? value), [defaultValue, value]);
+  const [isDragging, setIsDragging] = useState(false);
+  const [draggedIndex, setDraggedIndex] = useState(-1);
+
+  const markValues: Mark[] = useMemo(() => {
+    if (isArray(marks)) {
+      return marks.map(mark => ({...mark, value: validateValue(mark.value, min, max)}));
+    } else if (marks) {
+      return calculateMarks(min, max, validStep);
+    }
+    return [];
+  }, [marks, max, min, validStep]);
+
+  const tickMarks: Mark[] = useMemo(() => {
+    if (showTicks) {
+      return markValues.length ? markValues : calculateMarks(min, max, validStep);
+    }
+    return [];
+  }, [max, min, markValues, showTicks, validStep]);
+
+  const trackStart = useMemo(
+    () => toPercent(isRange ? Math.min(...validValues) : min, min, max),
+    [isRange, max, min, validValues]
+  );
+  const trackLength = useMemo(
+    () => toPercent(Math.max(...validValues), min, max) - trackStart,
+    [max, min, trackStart, validValues]
+  );
+
+  const handleValueChange = useCallback((nextValues: number[]) => {
+    setValues(nextValues);
+    onChange?.(isRange ? nextValues : nextValues[0]);
+  }, [isRange, onChange]);
+
+  const handleMouseDown = useCallback((e: React.MouseEvent) => {
+    e.stopPropagation();
+
+    if (disabled) {
+      return;
+    }
+
+    const index = e.currentTarget.getAttribute('data-index');
+    const nextValue = calculateValue(ref, e.pageX, min, max, validStep);
+    if (nextValue !== null && !isNaN(nextValue) && !index) {
+      const rangeIndex = Number(
+        Math.abs(validValues[0] - nextValue) > Math.abs(validValues[1] - nextValue)
+      );
+      setDraggedIndex(isRange ? rangeIndex : 0);
+    } else {
+      setDraggedIndex(Number(index));
+    }
+    setIsDragging(true);
+    previouslyDragged.current = false;
+  }, [disabled, isRange, max, min, validStep, validValues]);
+
+  const handleMouseUp = useCallback(({pageX}: MouseEvent) => {
+    const nextValues = adjustValues(validValues, ref, draggedIndex, pageX, max, min, validStep);
+    if (nextValues[0] > nextValues[1]) {
+      nextValues.reverse();
+    }
+    handleValueChange(nextValues);
+    setDraggedIndex(-1);
+    setIsDragging(false);
+    previouslyDragged.current = true;
+  }, [validValues, draggedIndex, handleValueChange, max, min, validStep]);
+
+  const handleMouseMove = useCallback(({pageX}: MouseEvent) => {
+    const nextValues = adjustValues(validValues, ref, draggedIndex, pageX, max, min, validStep);
+    if (nextValues[0] > nextValues[1]) {
+      nextValues.reverse();
+      setDraggedIndex(prevState => (prevState === 0 ? 1 : 0));
+    }
+    handleValueChange(nextValues);
+  }, [validValues, draggedIndex, max, min, validStep, handleValueChange]);
+
+  const handleKeyDown = useCallback(({key, currentTarget}: KeyboardEvent) => {
```

Comment:
I would request using `Shortcuts` [component](https://github.com/JetBrains/ring-ui/blob/master/src/shortcuts/shortcuts.ts) instead, because we try to implement shortcuts in a single way to make it unified and controlled

Reply:
Thank you for mentioning that, replaced it with Shortcuts